File: ExpectoRunner/SystemTestProvider.cs
using System;
using System.Linq;
using System.Reflection;
using System.Threading.Tasks;
using Mono.Addins;
using MonoDevelop.Core;
using MonoDevelop.Ide;
using MonoDevelop.Projects;

namespace MonoDevelop.UnitTesting.Expecto {
    public class SystemTestProvider : ITestProvider {
        public SystemTestProvider() { }

        public void Foo () {
            ITestProvider x = this;
            
        }

        public UnitTest CreateUnitTest(WorkspaceObject entry) {
            return new ExpectoTestSuite("My test");
        }

        public void Dispose() { }
    }
}

File: MonoDevelop.UnitTesting.Expecto/SystemTestProvider.fs
namespace MonoDevelop.UnitTesting.Expecto

open System
open MonoDevelop.UnitTesting
open MonoDevelop.Projects

type SystemTestProvider() =
    interface ITestProvider with
        member this.CreateUnitTest entry =
            match entry with
            | :? DotNetProject as project ->
                try
                    logfInfo "Creating test suite"
                    Some (new ExpectoProjectTestSuite(project) :> UnitTest)
                with e ->
                    logfError "Exception caught inside SystemTestProvider.CreateUnitTest: %A" e
                    None
            | _ -> None
            |> Option.toObj

        member this.Dispose () = ()